cleo_sdk/CLEO_Utils.h

```diff
@@ -31,9 +31,9 @@ namespace CLEO
     OPCODE_READ_PARAM_INT()
     OPCODE_READ_PARAM_UINT()
     OPCODE_READ_PARAM_FLOAT()
-    OPCODE_READ_PARAM_STRING() // returns char* to internal buffer. It might be overwritten by another string read!
+    OPCODE_READ_PARAM_STRING() // returns pointer to raw source data or CLEO's internal buffer, which MAY be overwritten by another OPCODE_READ_PARAM_STRING call
     OPCODE_READ_PARAM_STRING_BUFF(_buffer, _bufferSize) // always null terminated
-    OPCODE_READ_PARAM_FILEPATH() // returns char* to internal buffer. It might be overwritten by another string read!
+    OPCODE_READ_PARAM_FILEPATH() // returns pointer to internal buffer, which WILL be overwritten by another OPCODE_READ_PARAM_FILEPATH call
```

Comment:
oh boy, it's getting more and more complicated every day

Reply:
Actually it's comeback to old rules, only I finally understood why there were two separated exports for strings.
Plus more comments everywhere so no doubts for future readers.

Reply:
can you document all these rules?

Reply:
Already did. Both CLEO_ exports as well as OPCODE_READ_PARAM* have extensive comments.

Reply:
There are three methods:
```
CLEO_ReadStringOpcodeParam
CLEO_ReadStringPointerOpcodeParam
CLEO_ReadStringParamWriteBuffer
```
When should I use each of these?

Reply:
```
	//0ADF=2,add_dynamic_GXT_entry %1d% text %2d%
	OpcodeResult __stdcall opcode_0ADF(CRunningScript *thread)
	{
		char gxtBuff[8]; // 7 + terminator character
		auto gxt = OPCODE_READ_PARAM_STRING_BUFF(gxtBuff, sizeof(gxtBuff));
		auto txt = OPCODE_READ_PARAM_STRING();

		GetInstance().TextManager.AddFxt(gxt, txt);
		return OR_CONTINUE;
	}
```

why this opcode uses both `OPCODE_READ_PARAM_STRING_BUFF` and `OPCODE_READ_PARAM_STRING`? Why can't it use `OPCODE_READ_PARAM_STRING` twice?

Reply:
> There are three methods:
> 
> ```
> CLEO_ReadStringOpcodeParam
> CLEO_ReadStringPointerOpcodeParam
> CLEO_ReadStringParamWriteBuffer
> ```
> 
> When should I use each of these?

These things are literaly explained in CLEO.h next to the function
First two were in CLEO ealier.
CLEO_ReadStringParamWriteBuffer is used to store info about output string parameter, in cases when result in not last param of the opcode

Reply:
What script? I'm asking about opcode implementation. there is no script

Reply:
Depends on data provided by user in script. If raw data string uses no terminator then internal buffer must be utilized. Maybe it would be easier to say that buffer is always invalidated by second call.

Reply:
I think what you are saying, is that `OPCODE_READ_PARAM_STRING` uses shared buffer that will be overwritten when you call it multiple times in one opcode. which is why you need a combination of `OPCODE_READ_PARAM_STRING` and `OPCODE_READ_PARAM_STRING_BUFF` when opcode has multiple string arguments. Correct?

Reply:
Right, OPCODE_READ_PARAM_STRING has internal buffer (not always used).

Reply:
I wouldn't ask for example/documentation if those comments in cleo.h were clear to me. the fact that most opcodes don't use cleo.h directly but call utility macros from cleo_utils.h that have even more cryptic comments don't help at all. If I try to make my own opcode and learn by example, and look into existing opcodes implementation, how should I know about shared buffer if nothing tells me about it? I only found it in the implementation itself.

Examples of unclear comments:

> returns pointer to raw source data or CLEO's internal buffer, which MAY be overwritten by another OPCODE_READ_PARAM_STRING call

- will the returned pointer point to a null-terminated string? 
- no mention of the fact internal buffer is static (shared singleton)

```
LPSTR WINAPI CLEO_ReadStringOpcodeParam(CRunningScript* thread, char* buf = nullptr, int bufSize = 0); // writes null terminated string into buffer. If no buffer provided then internal one is used. Returns pointer to the result buffer or nullptr on fail
LPSTR WINAPI CLEO_ReadStringPointerOpcodeParam(CRunningScript* thread, char* buf = nullptr, int bufSize = 0); // writes null terminated string into buffer. Whenever possible returns pointer to the original, null terminated, source string (might be longer than provided buffer). Returns nullptr on fail.
```

- unclear when I should prefer one method over another

Reply:
Redacted comments a bit. Exports are difficult topic, as these were not documented at all.
Hard to say which should be used by default. Perhaps "Pointer" one. These should return **const** cstr, and "getPointer" makes no sense in the light of the function taking in buffer arguments.
Pretty sure people using it to write into strings, which works only in special cases.

Reply:
Question is, why don't we use `OPCODE_READ_PARAM_STRING_BUFF` for any string arguments and remove `OPCODE_READ_PARAM_STRING` version. I believe it would help us reduce the confusion and possible issues with values overwriting

Reply:
then we can get rid of a "internal buffer" thing and further simplify code (no nullptrs passed around)

Reply:
this is fine. define one buffer and read to it. no need to think twice

Reply:
I already rewrite all opcodes to use these macros and did not feel need for improvements.
`auto name = OPCODE_READ_PARAM_STRING();` is simplest thing you can imagine. Only rule - you can not call it twice. If you need it can also be made in simple way. As I did in copy file opcode, as it takes two paths.

If user is obligated to create buffer, then question emerges: what size that buffer should be? And if there is buffer we automatically need to copy each string into it every time. Buffer declarations won't make code look any better, as it starts introducing things that is not nature of the opcode itself. Read str, use it, set condition result, done.

Reply:
Buffer is always max_str size, except when we want to limit input, e.g. for a gxt label. We don't need to copy string there, if int value or literal fixed length string is given, we return a pointer without copying.

Biggest benefit would be reduced complexity of the code (no nullptr anywhere), also easier to reason about the code. No need to think twice when reading string argument. All that at a zero cost of an extra buffer declaration,

Reply:
> Only rule - you can not call it twice.

Best rule is when it is enforced with a code, not some comment.

Reply:
At this point you are trying to protect C++ developer from himself at expense of execution performance and code cleanness. Is somebody is unable to read single line of documentation or simply copy one of from dozens of the examples there is no more we can do for him.

Reply:
How are you want to solve it? When I call readString(buffer) I expect the buffer to contain the string. Not the string to be here sometime or sometime not, depending on script code. So it would require copying the string every time. 
That's what this hidden internal buffer is for. From user perspective he always receive pointer to the string with only the caution that the pointer is valid until next call of the macro.
You can then do std::string(ptr), or for more C way use read macro outputting to the provided buffer.

How exports work was already defined and can not be alerted as already proven it breaks existing mods.

Reply:
replace OPCODE_READ_PARAM_STRING with OPCODE_READ_PARAM_STRING_BUFF, then remove all buf==nullptr checks from _readParamText and `CLEO_ReadString...`

Reply:
providing implicit hidden buffer in cleo sdk was a mistake for the reasons above (you can't use it twice). if some opcodes/plugins rely on this, it's their fault. I quickly checked cleo+ and imgui plugins and they seem to provide their own buffers to each call which is good.

Reply:
Problem is the nullptr was supported by CLEO_ReadString... since ever, so no way to remove it now.
As mentioned above using OPCODE_READ_PARAM_STRING_BUFF, suggests the string will be in the buffer so it throws away possible optimization and requires always copy before use.
`_readParamText` is only used as wrapper for CLEO_ReadString... exports with additional checks.
None of `OPCODE_READ_PARAM_STRING` or `OPCODE_READ_PARAM_STRING_BUFF` expects nullptr args.

Reply:
If we get rid of `OPCODE_READ_PARAM_STRING` what is supposed `str = OPCODE_READ_PARAM_STRING_BUFF(buffer, sizeof(buffer);` to do then?

Reply:
> If we get rid of `OPCODE_READ_PARAM_STRING` what is supposed `str = OPCODE_READ_PARAM_STRING_BUFF(buffer, sizeof(buffer));` to do then?

same what it does now

Reply:
So there will be way to utilize fact that in half of scenarios cstring is already available to use. 
It always will be copied and arbitrary limited by buffer size.

Reply:
> So there will be way to utilize fact that in half of scenarios cstring is already available to use.
It always will be copied and arbitrary limited by buffer size.

can you give an example?

Reply:
strings can be provided from scripts in multiple types.
pointer to buffer is already cstring
other dynamic variables generally are null terminated
imm label and string also are in code terminated strings
exception are only CLEO's varstr and long\short dynamic variables filled to the full capacity.

Reply:
so what's the issue? you don't need to copy string to buffer if pointer was given, just return the pointer

Reply:
```
char* buf[max_str_len];
char* value = OPCODE_READ_PARAM_STRING_BUFF(buffer, sizeof(buffer));

// we don't care what happens in OPCODE_READ_PARAM_STRING_BUFF, but value will always be a pointer to a null-terminated string. we don't use `buf`
```

```

Reply:
> so what's the issue? you don't need to copy string to buffer if pointer was given, just return the pointer

```OPCODE_READ_PARAM_STRING_BUFF(buffer, sizeof(buffer));
print(buff);```
This is even worse trap than cstr invalidated by the next call.

Reply:
OPCODE_READ_PARAM_STRING_BUFF checks the argument. 

for int32 - it returns the value, no copy needed
for str8, str16 literal, its returns a pointer to them, no copy needed
string variable - copy to provided buffer, null-terminate , return pointer to this buffer
any other value - error, exit from opcode

Reply:
> > so what's the issue? you don't need to copy string to buffer if pointer was given, just return the pointer
> 
> ```
> OPCODE_READ_PARAM_STRING_BUFF(buffer, sizeof(buffer));
> print(buff);
> ```
> 
> This is even worse trap than cstr invalidated by the next call.

that's current behavior already. you shouldn't expect buf to change after the call.

Reply:
I mean something like this: https://github.com/cleolibrary/CLEO5/pull/79/files